Public/Src/Cache/ContentStore/Distributed/Sessions/DistributedContentSession.cs

```diff
@@ -283,42 +283,57 @@ private Task<ProactiveCopyResult> RequestProactiveCopyIfNeededAsync(OperationCon
                             return ProactiveCopyResult.CopyNotRequiredResult;
                         }
 
-                        IReadOnlyList<MachineLocation> buildRingMachines;
+                        IReadOnlyList<MachineLocation> buildRingMachines = null;
 
                         // Get random machine inside build ring
                         Task<BoolResult> insideRingCopyTask;
-                        if (_buildIdHash != null)
+                        if (Settings.EnableProactiveCopyInsideRing)

```

Comment:
Does it make sense to switch from having two bools to have "ProactiveCopyMode" with 3 values - Disable, InsideTheRingOnly, Full?

Reply:
If you'll stay with the current mode, we should assert that is impossible to "Enable" proactive copy but "disable" both - copy inside and outside the ring.

---
In reply to: [324778894](https://github.com/microsoft/BuildXL/pull/872#discussion_r324778894) [](ancestors = 324778894)

Reply:
I slightly prefer having the two configurations be independent of each other, so I'll add the assertion to make sure that we're on a valid state at all points.